crates/biome_js_analyze/src/semantic_analyzers/nursery/use_for_of.rs

```diff
@@ -0,0 +1,384 @@
+        };
+
+        if references.iter().all(index_only_used_with_array) {
+            Some(())
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(node: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.query().range(),
+            markup! {
+                "Use "<Emphasis>"for-of"</Emphasis>" loop instead of a "<Emphasis>"for loop"</Emphasis>"."
+            },
+        ))
+    }
+}
+
+/// Get initializer references taking scope to account (excluding initials `for` loop references)
+fn get_initializer_references(
+    block: Option<&biome_js_syntax::JsBlockStatement>,
+    binding: &biome_js_syntax::JsIdentifierBinding,
+    model: &biome_js_semantic::SemanticModel,
+) -> Vec<AnyJsExpression> {
+    let skip_references = {
+        if block.is_none() {
+            2
+        } else {
+            0
+        }
+    };
+
+    binding
+        .all_references(model)
+        .skip(skip_references)
+        .filter_map(|reference| {
+            if let Some(block) = block {
+                // If block is preset, we need to consider only references in this scope
+                if block.range().start() != reference.scope().range().start() {
+                    return None;
+                }
+            }
+
+            AnyJsExpression::cast(reference.syntax().parent()?)
+        })
+        .collect()
+}
+
+/// Ensure the `for` declaration is correctly initialized
+///
+/// ## Valid example
+/// for (let i = 0; i < array.length; i++)
+///
+/// ## Invalid examples
+/// for (let i = 0, x = 1; i < array.length; i++)
+///
+fn is_initializer_valid(initializer: &AnyJsForInitializer) -> Option<bool> {
+    let initializer_declarations = initializer.as_js_variable_declaration()?.declarators();
+    let initializer = initializer_declarations.first()?.ok()?;
+
+    if initializer_declarations.len() > 1 || !is_zero_initialized(&initializer)? {
+        return None;
+    }
+
+    Some(true)
+}
+
+/// Ensure the `for` test is correctly initialized
+///
+/// ## Valid example
+/// for (let i = 0; i < array.length; i++)
+///
+/// ## Invalid examples
+/// for (let i = 0; NOTI < array.length; i++)
+///
+fn is_test_valid(
+    initializer_binding: &JsIdentifierBinding,
+    test: &AnyJsExpression,
+) -> Option<bool> {
+    let test_binary_expression = test.as_js_binary_expression()?;
+    let left = test_binary_expression.left().ok()?;
+    let identifier_expression = left.as_js_identifier_expression()?;
+
+    if initializer_binding.name_token().ok()?.text_trimmed()
+        != identifier_expression
+            .name()
+            .ok()?
+            .value_token()
+            .ok()?
+            .text_trimmed()
+    {
+        return None;
+    }
+
+    if !is_less_than_length_expression(test_binary_expression)? {
+        return None;
+    }
+
+    Some(true)
+}
+
+/// Ensure the `for` update is correctly initialized
```

Comment:
The function name mentions the `update`, but the documentation talks about `for` Let's make sure to explain what this function does.